Add IPP plugin status tags to ZenDesk

iamgabrielma opened the pull request:
Closes: #7994 

### Description
With this PR, we'll be sending extra tags to ZenDesk that reflect the plugin status of `woocommerce-gateway-stripe` and `woocommerce-payments` plugins. There are 3 tags for each plugin:

- `_installed_and_activated`
- `_installed_and_not_activated`
- `_not_installed`

### Changes

In order to add these IPP statuses to ZenDesk tickets, we need to pass [these tags](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift#L625) through the SupportSDK's `RequestUiConfiguration` class when we create a new request. As we need to do this before ZenDesk tabs loads, we'll be using a `StorageSitePlugin` results controller directly through the `ZendeskManager` class, in order to fetch plugin statuses and pass these as tags.


### Testing instructions
The easiest way is through the debugger:
1. Confirm that you have [WooCommerce Payments](https://wordpress.org/plugins/woocommerce-payments/), and/or [WooCommerce Stripe Gateway](https://wordpress.org/plugins/woocommerce-gateway-stripe/) installed and/or activated
2. Add a breakpoint to [requestConfig.tags](https://github.com/woocommerce/woocommerce-ios/blob/b59a59d2efafc94dd861d15561cd1cac593da4b2/WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift#L631) to confirm which tags are we sending to ZenDesk.
3. Go to Menu > Settings > Help and Support > Tap on "Contact Support"
4. In Xcode's debugger, see the tags: 
```
tags	[String]	7 values	
[0]	String	"iOS"	
[1]	String	"woo-mobile-sdk"	
[2]	String	"jetpack"	
[3]	String	"woo_mobile_stripe_not_installed"	
[4]	String	"woo_mobile_wcpay_not_installed"	
[5]	String	"wpcom"	
[6]	String	"eCommerce"	
```
It can also be tested directly in ZenDesk, which requires extra steps:
1. Confirm that you have [WooCommerce Payments](https://wordpress.org/plugins/woocommerce-payments/), and/or [WooCommerce Stripe Gateway](https://wordpress.org/plugins/woocommerce-gateway-stripe/) installed and/or activated
2. Go to Menu > Settings > Help and Support > Confirm that you're logged into the app with a non-a8c account > Tap on "Contact Support" and send a request (You may need to test this on a physical device, not the simulator)
3. Log into ZenDesk, find your ticket, and confirm that the tags have been added and reflect the plugin status. If you have no ZD account, feel free to ping me on Slack with the non-a8c-email used (or any HE :D), in order to find the ticket.

<img width="355" alt="Screenshot 2022-11-07 at 12 30 43" src="https://user-images.githubusercontent.com/3812076/200220826-63d6f806-2ab5-4cc4-8cf5-0a11e5b4b841.png">

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

iamgabrielma committed:
Add getPluginDetails action to ZendeskManager

iamgabrielma committed:
Add plugin list and completion handler callback

iamgabrielma committed:
Try to get plugins by observing changes in Storage

iamgabrielma committed:
Append ipp tags if needed to decorated tags

iamgabrielma committed:
Add logic for plugin status

iamgabrielma committed:
Experiment with async call for tag retrieval

iamgabrielma committed:
Return all tags on creating ZD request

iamgabrielma committed:
Remove unused vars

iamgabrielma committed:
Move plugin slugs to Constants enum

iamgabrielma committed:
Access control changes. Make func private.

iamgabrielma committed:
Extract logic to SettingsViewModel

iamgabrielma committed:
Rename variables for clarity. Add comments.

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/Settings/SettingsViewModel.swift:
```diff
@@ -141,6 +155,34 @@ final class SettingsViewModel: SettingsViewModelOutput, SettingsViewModelActions
             stores.dispatch(action)
         }
+
+        ///
```
I forgot this internal comment. Will add or delete before merging.

iamgabrielma committed:
Move logic to ZendeskManager

We’re moving the logic to retrieve IPP plugin statuses directly to ZenDesk, as if a merchant attempts to open a ticket without passing through the SettingsView, we wouldn’t have these available.

iamgabrielma committed:
Remove unnecessary variable

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/Settings/SettingsViewModel.swift:
```diff
@@ -361,6 +361,17 @@ private extension SettingsViewModel {
 // MARK: - Localizations
 //
 private extension SettingsViewModel {
+    enum IPPPluginStatus {
+        static let stripe_plugin_slug = "woocommerce-gateway-stripe/woocommerce-gateway-stripe"
+        static let wcpay_plugin_slug = "woocommerce-payments/woocommerce-payments"
+        static let woo_mobile_stripe_not_installed = "woo_mobile_stripe_not_installed"
+        static let woo_mobile_stripe_installed_and_not_activated = "woo_mobile_stripe_installed_and_not_activated"
+        static let woo_mobile_stripe_installed_and_activated = "woo_mobile_stripe_installed_and_activated"
+        static let woo_mobile_wcpay_not_installed = "woo_mobile_wcpay_not_installed"
+        static let woo_mobile_wcpay_installed_and_not_activated = "woo_mobile_wcpay_installed_and_not_activated"
+        static let woo_mobile_wcpay_installed_and_activated = "woo_mobile_wcpay_installed_and_activated"
+        static let woo_mobile_site_plugins_fetching_error = "woo_mobile_site_plugins_fetching_error"
+    }
```
```suggestion
```
This needs to be removed as [is declared](https://github.com/woocommerce/woocommerce-ios/blob/9a1ae44d69d7f4bd509301cd3b40487d033a8f0a/WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift#L1120-L1130) on ZendeskManager now, I forgot 😬

iamgabrielma replied:
Fixed 849862e0e101050badcb6e53a96490ae3bc31dc8

itsmeichigo commented on WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift:
```diff
@@ -138,6 +137,40 @@ struct ZendeskProvider {
+
+    /// IPP plugin statuses
+    ///
+    private var ippPluginstatuses: [String]? {
```
Since `ippTags` is initialized with `[String]()` below, the default value would be an empty array. We should keep the return type as `[String]` then.

itsmeichigo commented on WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift:
```diff
@@ -324,7 +371,10 @@ final class ZendeskManager: NSObject, ZendeskManagerProtocol {
+        var tags = [Constants.platformTag, Constants.sdkTag, Constants.jetpackTag]
+
+        // Get IPP plugin statuses
+        tags.append(contentsOf: ippPluginstatuses ?? [])
```
If we change the type of `ippPluginstatuses` to non-optional, this can be simplified as:
```suggestion
        let tags = [Constants.platformTag, Constants.sdkTag, Constants.jetpackTag] + ippPluginstatuses
```

iamgabrielma replied:
That's right, we'll always be passing a value here so no reason for it to be optional. Changed here: c36468978777c5137a019484e90ea3e7ab561ed7

itsmeichigo commented on WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift:
```diff
@@ -1066,6 +1116,19 @@ private extension ZendeskManager {
+        static let woo_mobile_wcpay_installed_and_not_activated = "woo_mobile_wcpay_installed_and_not_activated"
+        static let woo_mobile_wcpay_installed_and_activated = "woo_mobile_wcpay_installed_and_activated"
+        static let woo_mobile_site_plugins_fetching_error = "woo_mobile_site_plugins_fetching_error"
+    }
```
Nit: I believe we prefer using camel case to snake case for variables, so for example, instead of `stripe_plugin_slug`, we should use `stripePluginSlug`.

Also, I see that you are having to repeatedly include the prefix `IPPPluginStatus` for each of these when defining `ippPluginstatuses`. My suggestion would be to separate the constants to different enums and make some of the static constants enum cases instead like so:

```suggestion
    enum PluginSlugs {
        static let stripe = "woocommerce-gateway-stripe/woocommerce-gateway-stripe"
        static let wcpay = "woocommerce-payments/woocommerce-payments"
     }
     enum PluginStatusTag: String {
        case stripeNotInstalled = "woo_mobile_stripe_not_installed"
        case stripeInstalledAndNotActivated = "woo_mobile_stripe_installed_and_not_activated"
        <# more cases here #>
    }
```

Then when defining `ippPluginstatuses` you can do as follows:

```Swift
    private var ippPluginstatuses: [String]? {
        var ippTags: [PluginStatusTag] = []
        if let stripe = pluginResultsController.fetchedObjects.first(where: { $0.plugin == PluginSlugs.stripe }) {
            if stripe.status == .active {
                ippTags.append(.stripeInstalledAndActivated)
            } else if stripe.status == .inactive {
                ippTags.append(.stripeInstalledAndNotActivated)
            }
        } 
        <# more conditions and checks here #>
        return ippTags.map { $0.rawValue }
    }
```

iamgabrielma replied:
Much better! Changed here 10fcac66cd9bf1beea2b6ed879fdb4a7b913ab87

itsmeichigo reviewed (commented):
I've tested the PR following the first method (adding a breakpoint to `tags`). I tried switching between different stores with different statuses for Stripe and WCPay. It looks like the tags are not correctly reflecting the plugin statuses when switching stores. 

Maybe we should synchronize plugins at some point to have the correct data in storage?

iamgabrielma committed:
Remove unused extension from SettingsViewModel

iamgabrielma committed:
Make enum cases snakecase and split them

iamgabrielma committed:
Make ippPluginstatuses non-optional

iamgabrielma committed:
Fix opening brace violation

iamgabrielma committed:
Move ResultsController outside of ZendeskManager init

As ZendeskManager is initialized only once, moving the ResultsController outside of the init allows for the siteID to be recalculated when we switch stores.

iamgabrielma committed:
Access ZD shared from SwitchStore

iamgabrielma committed:
Make pluginResultsController lazy

iamgabrielma commented:
> I've tested the PR following the first method (adding a breakpoint to tags). I tried switching between different stores with different statuses for Stripe and WCPay. It looks like the tags are not correctly reflecting the plugin statuses when switching stores. Maybe we should synchronize plugins at some point to have the correct data in storage?

Thanks for your review and help with this, this was definitely tricky. Fixed on e88a7a374ecb0363e8625e2ef99514979d829a52 . As we thought, we need to also call `pluginResultsController.performFetch()` in the initializer, otherwise will not work the first time we run it (only on store switcheroo).

Ready for review! 🙇

iamgabrielma committed:
Add missing method to MockZendeskManager

itsmeichigo commented on WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift:
```diff
@@ -138,6 +143,63 @@ struct ZendeskProvider {
+    private let stores = ServiceLocator.stores
+    private let storageManager = ServiceLocator.storageManager
+
+    /// Loads Plugins from the Storage Layer.
```
Super nit: 

```suggestion
    /// Controller for fetching site plugins from the storage
```

itsmeichigo commented on WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift:
```diff
@@ -138,6 +143,63 @@ struct ZendeskProvider {
+    ///
+    private lazy var pluginResultsController: ResultsController<StorageSitePlugin> = updatePluginResultsController()
+
+    private func updatePluginResultsController() -> ResultsController<StorageSitePlugin> {
```
Super nit:

```suggestion
    /// Returns a `pluginResultsController` using the latest selected site ID for predicate.
    private func createPluginResultsController() -> ResultsController<StorageSitePlugin> {
```

iamgabrielma replied:
I see! `create` makes more sense than `update` as we're not really updating it but creating a new one on each call. Changed on https://github.com/woocommerce/woocommerce-ios/pull/8008/commits/998499687560bf9c8621894c548b5e14208d4ba0

itsmeichigo commented on WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift:
```diff
@@ -138,6 +143,63 @@ struct ZendeskProvider {
+        }
+    }
+
+    /// IPP plugin statuses
```
Super nit:

```suggestion
    /// List of tags reflected statuses of Stripe and WCPay plugins
```

itsmeichigo commented on WooCommerce/Classes/Tools/Zendesk/ZendeskManager.swift:
```diff
@@ -138,6 +143,63 @@ struct ZendeskProvider {
+
+    /// IPP plugin statuses
+    ///
+    private var ippPluginstatuses: [String] {
```
Typo:

```suggestion
    private var ippPluginStatuses: [String] {
```

itsmeichigo reviewed (approved):
This works perfectly now 💯 I left a few more nits, please feel free to merge after updating them.

iamgabrielma committed:
Rename method for clarity. Typo.

iamgabrielma committed:
Remove trailing whitespace
